core/src/main/scala/org/apache/spark/sql/collection/ConcurrentSegmentedHashMap.scala

```diff
@@ -224,7 +225,10 @@ private[sql] class ConcurrentSegmentedHashMap[K, V, M <: SegmentMap[K, V] : Clas
             while (idx < nhashes) {
               added = seg.changeValue(keys(idx), bucketId(hashes(idx)), change, isLocal)
               if (added != null) {
-                if (added.booleanValue()) numAdded += 1
+                if (added.booleanValue()) {
+                  numAdded += 1
+                  this._size.incrementAndGet()
+                }
```

Comment:
This is taken care of by adding numValues at the end and numValue variable was precisely for that reason. This was kept so to avoid having to do a slightly expensive increment every time. An additional increment above will set the wrong size. This is for the non-region case only, right? For region case the size will be obtained from the bucket regions.

```scala
275:     }
276:     if (numAdded > 0) _size.addAndGet(numAdded)
```

The case that is probably not being handled is if there is an exception in between. Is that what you were looking at? For the exception case, it can be taken care of in the existing finally block:

```scala
            success = true
          } finally {
            if (lockedState) {
              lock.unlock()
            }
            if (!success) _size.addAndGet(numAdded)
          }
```